File: android/app/src/main/java/com/commuteoptimizer/widget/service/CommuteCalculator.kt
package com.commuteoptimizer.widget.service

import android.content.Context
import com.commuteoptimizer.widget.data.Result
import com.commuteoptimizer.widget.data.api.ApiClientFactory
import com.commuteoptimizer.widget.data.api.GoogleRoutesService
import com.commuteoptimizer.widget.data.api.MtaAlertsService
import com.commuteoptimizer.widget.data.api.MtaApiService
import com.commuteoptimizer.widget.data.models.*
import com.commuteoptimizer.widget.util.WidgetPreferences
import java.text.SimpleDateFormat
import java.util.*

/**
 * Main calculator that orchestrates API calls and builds commute options.
 * Uses MTA GTFS-realtime API directly (no rate limits).
 * Optionally uses Google Routes API via Cloudflare Worker for full transit routes.
 */
class CommuteCalculator(private val context: Context) {

    private val prefs = WidgetPreferences(context)
    private val weatherApi = ApiClientFactory.weatherApi
    private val localDataSource = LocalDataSource(context)

    suspend fun calculateCommute(widgetId: Int = -1): Result<CommuteResponse> {
        try {
            // Use per-widget origin if widgetId is provided, otherwise fall back to global
            val homeLat = if (widgetId >= 0) prefs.getWidgetOriginLat(widgetId) else prefs.getHomeLat()
            val homeLng = if (widgetId >= 0) prefs.getWidgetOriginLng(widgetId) else prefs.getHomeLng()
            // Use per-widget destination if widgetId is provided, otherwise fall back to global
            val workLat = if (widgetId >= 0) prefs.getWidgetDestinationLat(widgetId) else prefs.getWorkLat()
            val workLng = if (widgetId >= 0) prefs.getWidgetDestinationLng(widgetId) else prefs.getWorkLng()
            val apiKey = prefs.getOpenWeatherApiKey()
            val googleApiKey = prefs.getGoogleApiKey()
            val showBikeOptions = prefs.getShowBikeOptions()

            if (homeLat == 0.0 || workLat == 0.0) {
                return Result.Error("Please configure home and work locations in settings")
            }

            val autoSelected = localDataSource.autoSelectStations(homeLat, homeLng, workLat, workLng)
            if (autoSelected.isEmpty()) {
                return Result.Error("No stations found within 4 miles of home")
            }

            val weather = fetchWeather(homeLat, homeLng, apiKey)
            val stations = localDataSource.getStations()
            val stationMap = stations.associateBy { it.id }

            val destStation = findClosestStation(workLat, workLng, stations)
                ?: return Result.Error("No station found near work location")

            val options = mutableListOf<CommuteOption>()

            // Add walk-only option if work is close
            val homeToWorkDist = DistanceCalculator.haversineDistance(homeLat, homeLng, workLat, workLng)
            if (homeToWorkDist < 2.0) {
                options.add(buildWalkOnlyOption(homeLat, homeLng, workLat, workLng))
            }

            // Build bike-to-transit options for all selected stations
            for ((index, stationId) in autoSelected.withIndex()) {
                val station = stationMap[stationId] ?: continue

                try {
                    if (showBikeOptions) {
                        buildBikeToTransitOption(
                            station, homeLat, homeLng, destStation, workLat, workLng,
                            googleApiKey, index
                        )?.let { options.add(it) }
                    }
                } catch (e: Exception) {
                    continue
                }
            }

            // Build transit-only (walk) options for top 3 closest stations (like webapp)
            val walkableStations = autoSelected
                .mapNotNull { stationId -> stationMap[stationId]?.let { stationId to it } }
                .map { (stationId, station) ->
                    val walkTime = DistanceCalculator.estimateWalkTime(homeLat, homeLng, station.lat, station.lng)
                    Triple(stationId, station, walkTime)
                }
                .sortedBy { it.third } // Sort by walk time
                .take(3) // Top 3 closest

            for ((index, triple) in walkableStations.withIndex()) {
                val (stationId, station, walkTime) = triple
                try {
                    buildTransitOnlyOption(
                        station, homeLat, homeLng, destStation, workLat, workLng,
                        googleApiKey, walkTime, index + 100
                    )?.let { options.add(it) }
                } catch (e: Exception) {
                    continue
                }
            }

            if (options.isEmpty()) {
                return Result.Error("No commute options available")
            }

            // Deduplicate options by transit route signature (keep fastest for each unique route)
            val deduplicated = deduplicateByRoute(options)

            val ranked = RankingService.rankOptions(deduplicated, weather)

            val routeIds = autoSelected.flatMap { id ->
                stationMap[id]?.lines ?: emptyList()
            }.distinct()
            val alerts = fetchAlerts(routeIds)

            val response = CommuteResponse(
                options = ranked.take(3),
                weather = weather,
                alerts = alerts,
                generatedAt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
                    .apply { timeZone = TimeZone.getTimeZone("UTC") }
                    .format(Date())
            )

            return Result.Success(response)

        } catch (e: Exception) {
            return Result.Error("Failed to calculate commute: ${e.message}", e)
        }
    }

    private fun findClosestStation(lat: Double, lng: Double, stations: List<LocalStation>): LocalStation? {
        return stations.minByOrNull { station ->
            DistanceCalculator.haversineDistance(lat, lng, station.lat, station.lng)
        }
    }

    private suspend fun fetchWeather(lat: Double, lng: Double, apiKey: String?): Weather {
        if (apiKey.isNullOrBlank()) {
            android.util.Log.w("CommuteCalc", "No OpenWeather API key configured")
            return getDefaultWeather()
        }

        return try {
            android.util.Log.d("CommuteCalc", "Fetching weather for $lat, $lng")
            val response = weatherApi.getWeather(lat, lng, apiKey = apiKey)
            if (response.isSuccessful && response.body() != null) {
                android.util.Log.d("CommuteCalc", "Weather fetched successfully: ${response.body()?.main?.temp}")
                parseWeatherResponse(response.body()!!)
            } else {
                android.util.Log.e("CommuteCalc", "Weather API error: ${response.code()} - ${response.errorBody()?.string()}")
                getDefaultWeather()
            }
        } catch (e: Exception) {
            android.util.Log.e("CommuteCalc", "Weather fetch exception: ${e.message}", e)
            getDefaultWeather()
        }
    }

    private fun parseWeatherResponse(data: OpenWeatherResponse): Weather {
        val weatherId = data.weather.firstOrNull()?.id ?: 800
        val weatherMain = data.weather.firstOrNull()?.main ?: "Clear"

        val precipitationType = when {
            weatherId in 200..599 -> "rain"
            weatherId in 600..610 -> "snow"
            weatherId in 611..699 -> "mix"
            else -> "none"
        }

        // Check if there's active rain or snow from the response
        val hasRain = (data.rain?.oneHour ?: data.rain?.threeHour ?: 0.0) > 0
        val hasSnow = (data.snow?.oneHour ?: data.snow?.threeHour ?: 0.0) > 0
        val isBad = precipitationType != "none" || hasRain || hasSnow

        return Weather(
            tempF = data.main.temp.toInt(),
            conditions = weatherMain,
            precipitationType = precipitationType,
            precipitationProbability = if (isBad) 100 else 0,
            isBad = isBad
        )
    }

    private fun getDefaultWeather() = Weather(null, "Unknown", "none", 0, false)

    private suspend fun fetchAlerts(routeIds: List<String>): List<Alert> {
        return try {
            MtaAlertsService.fetchAlerts(routeIds)
                .filter { it.effect in listOf("NO_SERVICE", "REDUCED_SERVICE", "SIGNIFICANT_DELAYS") }
                .take(3)
                .map { Alert(it.routeIds, it.effect, it.headerText) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun buildWalkOnlyOption(homeLat: Double, homeLng: Double, workLat: Double, workLng: Double): CommuteOption {
        val walkTime = DistanceCalculator.estimateWalkTime(homeLat, homeLng, workLat, workLng)
        val arrivalTime = Calendar.getInstance().apply { add(Calendar.MINUTE, walkTime) }
        val arrivalTimeStr = SimpleDateFormat("h:mm a", Locale.US).format(arrivalTime.time)

        return CommuteOption(
            id = "walk_only", rank = 0, type = "walk_only",
            durationMinutes = walkTime, summary = "Walk to Work",
            legs = listOf(Leg(mode = "walk", duration = walkTime, to = "Work", route = null, from = "Home")),
            nextTrain = "N/A", arrivalTime = arrivalTimeStr,
            station = Station("", "Work", "", emptyList(), workLat, workLng, "")
        )
    }

    private suspend fun buildBikeToTransitOption(
        station: LocalStation, homeLat: Double, homeLng: Double,
        destStation: LocalStation, workLat: Double, workLng: Double,
        googleApiKey: String?, index: Int
    ): CommuteOption? {
        val bikeTime = DistanceCalculator.estimateBikeTime(homeLat, homeLng, station.lat, station.lng)
        val nextArrival = getNextArrival(station.mtaId, station.lines)
        val transitRoute = getTransitRoute(station, destStation, workLat, workLng, googleApiKey)
        if (transitRoute == null) return null
        val (transitTime, transitLegs) = transitRoute

        // Include wait time like webapp: bikeTime + waitTime + transitTime
        val waitTime = nextArrival.minutesAway.coerceAtLeast(0)
        val totalDuration = bikeTime + waitTime + transitTime
        val legs = mutableListOf(Leg(mode = "bike", duration = bikeTime, to = station.name, route = null, from = "Home"))
        legs.addAll(transitLegs)

        // Build summary with final stop like webapp: "Bike → G → Court Sq"
        val summary = buildSummary("Bike", transitLegs, nextArrival.routeId ?: station.lines.firstOrNull(), destStation.name)
        val arrivalTime = Calendar.getInstance().apply { add(Calendar.MINUTE, totalDuration) }
        val arrivalTimeStr = SimpleDateFormat("h:mm a", Locale.US).format(arrivalTime.time)

        return CommuteOption(
            id = "bike_$index", rank = 0, type = "bike_to_transit",
            durationMinutes = totalDuration, summary = summary, legs = legs,
            nextTrain = nextArrival.nextTrainText, arrivalTime = arrivalTimeStr,
            station = Station(station.id, station.name, station.mtaId, station.lines, station.lat, station.lng, station.borough)
        )
    }

    private suspend fun buildTransitOnlyOption(
        station: LocalStation, homeLat: Double, homeLng: Double,
        destStation: LocalStation, workLat: Double, workLng: Double,
        googleApiKey: String?, walkTime: Int, index: Int
    ): CommuteOption? {
        val nextArrival = getNextArrival(station.mtaId, station.lines)
        val transitRoute = getTransitRoute(station, destStation, workLat, workLng, googleApiKey)
        if (transitRoute == null) return null
        val (transitTime, transitLegs) = transitRoute

        // Include wait time like webapp: walkTime + waitTime + transitTime
        val waitTime = nextArrival.minutesAway.coerceAtLeast(0)
        val totalDuration = walkTime + waitTime + transitTime
        val legs = mutableListOf(Leg(mode = "walk", duration = walkTime, to = station.name, route = null, from = "Home"))
        legs.addAll(transitLegs)

        // Build summary with final stop like webapp: "Walk → G → Court Sq"
        val summary = buildSummary("Walk", transitLegs, nextArrival.routeId ?: station.lines.firstOrNull(), destStation.name)
        val arrivalTime = Calendar.getInstance().apply { add(Calendar.MINUTE, totalDuration) }
        val arrivalTimeStr = SimpleDateFormat("h:mm a", Locale.US).format(arrivalTime.time)

        return CommuteOption(
            id = "transit_$index", rank = 0, type = "transit_only",
            durationMinutes = totalDuration, summary = summary, legs = legs,
            nextTrain = nextArrival.nextTrainText, arrivalTime = arrivalTimeStr,
            station = Station(station.id, station.name, station.mtaId, station.lines, station.lat, station.lng, station.borough)
        )
    }

    private suspend fun getTransitRoute(
        fromStation: LocalStation, toStation: LocalStation,
        workLat: Double, workLng: Double,
        googleApiKey: String?
    ): Pair<Int, List<Leg>>? {
        if (!googleApiKey.isNullOrBlank()) {
            try {
                val result = GoogleRoutesService.getTransitRoute(
                    googleApiKey, fromStation.lat, fromStation.lng, workLat, workLng
                )
                if (result.status == "OK" && result.durationMinutes != null) {
                    val legs = result.transitSteps.map { step ->
                        Leg(
                            mode = "subway",
                            duration = step.duration ?: 0,
                            to = step.arrivalStop ?: "?",
                            route = step.line,
                            from = step.departureStop,
                            numStops = step.numStops
                        )
                    }
                    return Pair(result.durationMinutes, legs.ifEmpty {
                        listOf(Leg("subway", result.durationMinutes, toStation.name, fromStation.lines.firstOrNull(), fromStation.name, null))
                    })
                }
            } catch (e: Exception) { }
        }

        // Google Routes API required but failed or not configured
        return null
    }

    private fun buildSummary(firstMode: String, transitLegs: List<Leg>, firstLine: String?, finalStop: String): String {
        val lines = transitLegs.mapNotNull { it.route }.distinct()
        val linesSummary = if (lines.isNotEmpty()) lines.joinToString(" → ") else (firstLine ?: "?")
        // Get final destination from last leg or use provided finalStop
        val destination = transitLegs.lastOrNull()?.to ?: finalStop
        return "$firstMode → $linesSummary → $destination"
    }

    data class NextArrival(
        val nextTrainText: String,
        val arrivalTime: String,
        val routeId: String?,
        val minutesAway: Int
    )

    private suspend fun getNextArrival(stationId: String, lines: List<String>): NextArrival {
        return try {
            val arrivals = MtaApiService.getStationArrivals(stationId, lines)
            if (arrivals.isEmpty()) {
                NextArrival("--", "--", null, 5) // Default 5 min wait if no data
            } else {
                val next = arrivals.first()
                val arrivalDate = Date(next.arrivalTime * 1000)
                val timeFormat = SimpleDateFormat("h:mm a", Locale.US)
                val displayText = if (next.minutesAway <= 0) "Now" else "${next.minutesAway}m"
                NextArrival(
                    nextTrainText = displayText,
                    arrivalTime = timeFormat.format(arrivalDate),
                    routeId = next.routeId,
                    minutesAway = next.minutesAway
                )
            }
        } catch (e: Exception) {
            NextArrival("--", "--", null, 5)
        }
    }

    /**
     * Deduplicate commute options by transit route signature.
     * If multiple options use the same sequence of subway lines, keep only the fastest.
     * This prevents showing "Bike → Station A → G" and "Bike → Station B → G" separately.
     */
    private fun deduplicateByRoute(options: List<CommuteOption>): List<CommuteOption> {
        val bySignature = mutableMapOf<String, CommuteOption>()

        for (option in options) {
            // Build route signature: type + sequence of subway lines
            val subwayLegs = option.legs.filter { it.mode == "subway" && it.route != null }
            val routeSequence = subwayLegs.mapNotNull { it.route }.joinToString("→")
            val signature = "${option.type}_$routeSequence"

            val existing = bySignature[signature]
            if (existing == null || option.durationMinutes < existing.durationMinutes) {
                bySignature[signature] = option
            }
        }

        return bySignature.values.toList()
    }
}

File: ios/CommuteOptimizer/CommuteOptimizer/Services/CommuteCalculator.swift
import Foundation

actor CommuteCalculator {
    private let weatherService = WeatherApiService()
    private let mtaService = MtaApiService()
    private let mtaAlertsService = MtaAlertsService()
    private let googleRoutesService = GoogleRoutesService()
    private let stationsDataSource = StationsDataSource.shared

    func calculateCommute(settings: SettingsManager) async throws -> CommuteResponse {
        try await calculateCommute(
            settings: settings,
            originLat: settings.homeLat,
            originLng: settings.homeLng,
            destLat: settings.workLat,
            destLng: settings.workLng
        )
    }

    /// Calculate commute with custom origin/destination (for per-widget configuration)
    func calculateCommute(
        settings: SettingsManager,
        originLat: Double,
        originLng: Double,
        destLat: Double,
        destLng: Double
    ) async throws -> CommuteResponse {
        let stations = stationsDataSource.getStations()

        // 1. Fetch weather at origin
        let weather = await fetchWeather(lat: originLat, lng: originLng, apiKey: settings.openWeatherApiKey)

        // 2. Find destination station (closest to destination)
        guard let destStation = findClosestStation(
            toLat: destLat,
            toLng: destLng,
            from: stations
        ) else {
            throw CommuteError.noDestinationStation
        }

        var options: [CommuteOption] = []

        // 3. Walk-only option (if < 2 miles)
        let homeToWorkDistance = DistanceCalculator.haversineDistance(
            lat1: originLat, lon1: originLng,
            lat2: destLat, lon2: destLng
        )

        if homeToWorkDistance < 2 {
            let walkTime = DistanceCalculator.estimateWalkTime(
                fromLat: originLat, fromLng: originLng,
                toLat: destLat, toLng: destLng
            )

            let arrivalDate = Date().addingTimeInterval(TimeInterval(walkTime * 60))
            let formatter = DateFormatter()
            formatter.dateFormat = "h:mm a"

            options.append(CommuteOption(
                id: "walk-only",
                rank: 0,
                type: .walkOnly,
                durationMinutes: walkTime,
                summary: "Walk to Work",
                legs: [Leg(mode: .walk, duration: walkTime, to: "Work")],
                nextTrain: "N/A",
                arrivalTime: formatter.string(from: arrivalDate),
                station: Station(
                    id: destStation.id,
                    name: destStation.name,
                    mtaId: destStation.mtaId,
                    lines: destStation.lines,
                    lat: destStation.lat,
                    lng: destStation.lng,
                    borough: destStation.borough
                )
            ))
        }

        // 4. Bike-to-transit options (auto-selected stations)
        let autoSelected = stationsDataSource.autoSelectStations(
            homeLat: originLat, homeLng: originLng,
            workLat: destLat, workLng: destLng
        )

        if settings.showBikeOptions {
            for stationId in autoSelected {
                if let option = await buildBikeToTransitOption(
                    stationId: stationId,
                    destStation: destStation,
                    originLat: originLat,
                    originLng: originLng,
                    destLat: destLat,
                    destLng: destLng,
                    settings: settings
                ) {
                    options.append(option)
                }
            }
        }

        // 5. Transit-only options (top 3 closest from auto-selected)
        let autoStations = autoSelected.compactMap { stationsDataSource.getStation(id: $0) }
        let walkSorted = autoStations.sorted {
            DistanceCalculator.estimateWalkTime(fromLat: originLat, fromLng: originLng, toLat: $0.lat, toLng: $0.lng) <
            DistanceCalculator.estimateWalkTime(fromLat: originLat, fromLng: originLng, toLat: $1.lat, toLng: $1.lng)
        }

        for station in walkSorted.prefix(3) {
            if let option = await buildTransitOnlyOption(
                station: station,
                destStation: destStation,
                originLat: originLat,
                originLng: originLng,
                destLat: destLat,
                destLng: destLng,
                settings: settings
            ) {
                options.append(option)
            }
        }

        // 6. Deduplicate options by route signature
        let deduplicated = deduplicateOptions(options)

        // 7. Rank options
        let rankedOptions = RankingService.rankOptions(deduplicated, weather: weather)

        // 8. Fetch alerts
        let routeIds = Set(rankedOptions.flatMap { $0.station.lines })
        let alerts = await fetchAlerts(routeIds: Array(routeIds))

        // 9. Return top 3
        let formatter = ISO8601DateFormatter()
        return CommuteResponse(
            options: Array(rankedOptions.prefix(3)),
            weather: weather,
            alerts: alerts,
            generatedAt: formatter.string(from: Date())
        )
    }

    // MARK: - Private Methods

    private func fetchWeather(settings: SettingsManager) async -> Weather {
        await fetchWeather(lat: settings.homeLat, lng: settings.homeLng, apiKey: settings.openWeatherApiKey)
    }

    private func fetchWeather(lat: Double, lng: Double, apiKey: String) async -> Weather {
        do {
            return try await weatherService.getWeather(
                lat: lat,
                lng: lng,
                apiKey: apiKey
            )
        } catch {
            return Weather(tempF: 65, conditions: "Unknown", precipitationType: .none, precipitationProbability: 0, isBad: false)
        }
    }

    private func findClosestStation(toLat: Double, toLng: Double, from stations: [LocalStation]) -> LocalStation? {
        stations.min { a, b in
            let distA = DistanceCalculator.haversineDistance(lat1: toLat, lon1: toLng, lat2: a.lat, lon2: a.lng)
            let distB = DistanceCalculator.haversineDistance(lat1: toLat, lon1: toLng, lat2: b.lat, lon2: b.lng)
            return distA < distB
        }
    }

    /// Deduplicate options by route signature to prevent showing same transit route from multiple bike stations
    private func deduplicateOptions(_ options: [CommuteOption]) -> [CommuteOption] {
        var seen: [String: CommuteOption] = [:]

        for option in options {
            // Build route signature: "bike_to_transit_G" or "transit_only_A->6"
            let routeSequence = option.legs
                .filter { $0.mode == .subway }
                .compactMap { $0.route }
                .joined(separator: "->")
            let signature = "\(option.type.rawValue)_\(routeSequence)"

            // Keep fastest option for each signature
            if let existing = seen[signature] {
                if option.durationMinutes < existing.durationMinutes {
                    seen[signature] = option
                }
            } else {
                seen[signature] = option
            }
        }

        return Array(seen.values)
    }

    /// Get transit route using Google API if available, fallback to local estimate
    private func getTransitRoute(
        fromStation: LocalStation,
        toStation: LocalStation,
        workLat: Double,
        workLng: Double,
        googleApiKey: String?
    ) async -> (duration: Int, legs: [Leg]) {
        // Try Google Routes API if key is provided
        if let apiKey = googleApiKey, !apiKey.isEmpty {
            let result = await googleRoutesService.getTransitRoute(
                apiKey: apiKey,
                originLat: fromStation.lat, originLng: fromStation.lng,
                destLat: workLat, destLng: workLng
            )

            if result.status == "OK", let duration = result.durationMinutes {
                let legs = result.transitSteps.map { step in
                    Leg(
                        mode: .subway,
                        duration: step.duration ?? 0,
                        to: step.arrivalStop ?? "?",
                        route: step.line,
                        from: step.departureStop,
                        numStops: step.numStops
                    )
                }

                if !legs.isEmpty {
                    return (duration, legs)
                }

                // Google returned duration but no detailed steps
                return (duration, [Leg(mode: .subway, duration: duration, to: toStation.name, route: fromStation.lines.first, from: fromStation.name)])
            }
        }

        // Fallback to local estimate
        let transitTime = DistanceCalculator.estimateTransitTime(
            fromStopId: fromStation.mtaId,
            toStopId: toStation.mtaId
        )
        return (transitTime, [Leg(mode: .subway, duration: transitTime, to: toStation.name, route: fromStation.lines.first, from: fromStation.name)])
    }

    private func buildBikeToTransitOption(
        stationId: String,
        destStation: LocalStation,
        originLat: Double,
        originLng: Double,
        destLat: Double,
        destLng: Double,
        settings: SettingsManager
    ) async -> CommuteOption? {
        guard let station = stationsDataSource.getStation(id: stationId) else { return nil }

        let bikeTime = DistanceCalculator.estimateBikeTime(
            fromLat: originLat, fromLng: originLng,
            toLat: station.lat, toLng: station.lng
        )

        let arrival = await mtaService.getNextArrival(stationId: station.mtaId, lines: station.lines)
        let waitTime = arrival.minutesAway

        // Get transit route (uses Google API if available)
        let (transitTime, transitLegs) = await getTransitRoute(
            fromStation: station,
            toStation: destStation,
            workLat: destLat,
            workLng: destLng,
            googleApiKey: settings.googleApiKey
        )

        let totalDuration = bikeTime + waitTime + transitTime
        let route = arrival.routeId ?? transitLegs.first?.route ?? station.lines.first ?? "?"
        let finalStop = transitLegs.last?.to ?? destStation.name

        let arrivalDate = Date().addingTimeInterval(TimeInterval(totalDuration * 60))
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"

        // Build summary like Android: "Bike -> G -> Court Sq" or "Bike -> A -> G -> Court Sq"
        let summary = buildSummary(firstMode: "Bike", transitLegs: transitLegs, firstLine: route, finalStop: finalStop)

        return CommuteOption(
            id: "bike-\(stationId)",
            rank: 0,
            type: .bikeToTransit,
            durationMinutes: totalDuration,
            summary: summary,
            legs: [Leg(mode: .bike, duration: bikeTime, to: station.name)] + transitLegs,
            nextTrain: arrival.nextTrain,
            arrivalTime: formatter.string(from: arrivalDate),
            station: Station(
                id: station.id,
                name: station.name,
                mtaId: station.mtaId,
                lines: station.lines,
                lat: station.lat,
                lng: station.lng,
                borough: station.borough
            )
        )
    }

    private func buildTransitOnlyOption(
        station: LocalStation,
        destStation: LocalStation,
        originLat: Double,
        originLng: Double,
        destLat: Double,
        destLng: Double,
        settings: SettingsManager
    ) async -> CommuteOption? {
        let walkTime = DistanceCalculator.estimateWalkTime(
            fromLat: originLat, fromLng: originLng,
            toLat: station.lat, toLng: station.lng
        )

        let arrival = await mtaService.getNextArrival(stationId: station.mtaId, lines: station.lines)
        let waitTime = arrival.minutesAway

        // Get transit route (uses Google API if available)
        let (transitTime, transitLegs) = await getTransitRoute(
            fromStation: station,
            toStation: destStation,
            workLat: destLat,
            workLng: destLng,
            googleApiKey: settings.googleApiKey
        )

        let totalDuration = walkTime + waitTime + transitTime
        let route = arrival.routeId ?? transitLegs.first?.route ?? station.lines.first ?? "?"
        let finalStop = transitLegs.last?.to ?? destStation.name

        let arrivalDate = Date().addingTimeInterval(TimeInterval(totalDuration * 60))
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"

        // Build summary like Android: "Walk -> G -> Court Sq" or "Walk -> A -> G -> Court Sq"
        let summary = buildSummary(firstMode: "Walk", transitLegs: transitLegs, firstLine: route, finalStop: finalStop)

        return CommuteOption(
            id: "transit-\(station.id)",
            rank: 0,
            type: .transitOnly,
            durationMinutes: totalDuration,
            summary: summary,
            legs: [Leg(mode: .walk, duration: walkTime, to: station.name)] + transitLegs,
            nextTrain: arrival.nextTrain,
            arrivalTime: formatter.string(from: arrivalDate),
            station: Station(
                id: station.id,
                name: station.name,
                mtaId: station.mtaId,
                lines: station.lines,
                lat: station.lat,
                lng: station.lng,
                borough: station.borough
            )
        )
    }

    /// Build summary string like Android: "Mode -> Lines -> FinalStop"
    private func buildSummary(firstMode: String, transitLegs: [Leg], firstLine: String?, finalStop: String) -> String {
        let lines = transitLegs.compactMap { $0.route }.filter { !$0.isEmpty }
        let uniqueLines = lines.isEmpty ? [firstLine ?? "?"] : Array(NSOrderedSet(array: lines)) as! [String]
        let linesSummary = uniqueLines.joined(separator: " -> ")
        let destination = transitLegs.last?.to ?? finalStop
        return "\(firstMode) -> \(linesSummary) -> \(destination)"
    }

    private func fetchAlerts(routeIds: [String]) async -> [ServiceAlert] {
        let alerts = await mtaAlertsService.fetchAlerts(routeIds: routeIds)
        // Filter to only show significant alerts and limit to 3
        return Array(alerts.filter {
            ["NO_SERVICE", "REDUCED_SERVICE", "SIGNIFICANT_DELAYS"].contains($0.effect)
        }.prefix(3))
    }
}

enum CommuteError: Error {
    case noDestinationStation
    case notConfigured
}
